blueprints/hr_assistant/general.py

```diff
@@ -0,0 +1,640 @@
+            gte_val = np.min(num_entity)
+            lte_val = np.max(num_entity)
+
+        qa = qa.filter(field='age', gte=gte_val, lte=lte_val)
+
+    elif len(num_entity) >= 1:
+        qa = qa.filter(field='age', gte=np.min(num_entity), lte=np.max(num_entity))
+
+    else:
+        size = SIZE
+
+    return qa, size
+
+
+def _find_additional_age_entities(request, responder):
+    """
+    If the user has a query such as 'list all employees under 30', the notion of age is
+    implicit rather than explicit in the form of an age entity. Hence, this function is
+    beneficial in capturing the existence such implicit entities.
+
+    Returns a true/false depending on the existence or lack of the combination of
+    numerical entities and comparators, thereby indicating an implicit age entitiy or
+    lack of it, respectively.
+    """
+    try:
+        comparator_entity = [e for e in request.entities if e['type'] == 'comparator'][0]
+        num_entity = [float(e['value'][0]['value'])
+                      for e in request.entities
+                      if e['type'] == 'sys_number']
+
+        # if any token in the text query is numeric that was missed by the num_entity,
+        # add it to the list
+        for i in request.text.split():
+            try:
+                num_entity.append(float(i))
+            except Exception:
+                continue
+    except Exception:
+        comparator_entity = []
+        num_entity = []
+
+    return True if comparator_entity and num_entity else False
+
+
+def _resolve_categorical_entities(request, responder):
+    """
+    This function retrieves all categorical entities as listed below and filters
+    the knowledge base using these entities as filters. The final search object
+    containing the shortlisted employee data is returned back to the calling function.
+    """
+
+    # Finding all categorical entities
+    categorical_entities = [e for e in request.entities if e['type'] in
+                            ('state', 'sex', 'maritaldesc', 'citizendesc', 'racedesc',
+                             'performance_score', 'employment_status', 'employee_source',
+                             'position', 'department', 'reason_for_termination')]
+
+    # Building custom search
+    qa = app.question_answerer.build_search(index='employee_data')
+
+    # Querying the knowledge base for all categorical filters
+    if categorical_entities:
+        try:
+            for categorical_entity in categorical_entities:
+                key = categorical_entity['type']
+
+                if key == 'reason_for_termination':
+                    key = 'rft'
+
+                val = categorical_entity['value'][0]['cname']
+                kw = {key: val}
+                qa = qa.filter(**kw)
+        except Exception:
+            pass
+
+    # return size of the whole dataset to prevent the execute function from restricting
+    # the responses to 10 (Which is the default)
+    size = SIZE
```

Comment:
size is not used in this function, maybe it shouldn't be a part of this function